### PR TITLE
logging: log: add `LOG_WRN_ONCE`

### PR DESCRIPTION
**Pull request by JordanYates**
Add an equivalent to the linux `WARN_ONCE` macro. This is intended for
use when the developer should be notified of an event, but may occur a
multitude of times in quick succession.

Using `LOG_WRN` could result in either cluttered logs or recursive
logging (i.e. in serial drivers).

Proposed in the discussion of #65112

### CODE REVIEW

File: include/zephyr/logging/log.h [hunk 1]
```diff
@@ -72,6 +72,24 @@ extern "C" {
...
+ *
+ * @details It's meant for situations that warrant investigation but could clutter
+ * the logs if output on every execution.
+ *
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_WRN_ONCE(...)					\
+	do {							\
+		static int __warned;				\
```
Comment (tmon-nordic) [inline include/zephyr/logging/log.h:86]: What about using some smaller data type?

Comment (JordanYates) [reply include/zephyr/logging/log.h]: I was trying to avoid pulling in other headers but I see `<stdint.h>` types are already being used.

**Review by tmon-nordic — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]
LGTM - I tossed around a similar rate-limit debug variant at one point.

**Comment by nordic-krch** [comment]
Long time ago there was #41783 but author dropped it and it never get in. There we had `LOG_<lvl>_N(n, ...)` which was more flexible and supported all severity levels maybe it would be good to resurrect that approach instead?

**Comment by JordanYates** [comment]
> Long time ago there was #41783 but author dropped it and it never get in. There we had `LOG_<lvl>_N(n, ...)` which was more flexible and supported all severity levels maybe it would be good to resurrect that approach instead?

I don't mind if that is the approach you want to go with, but I won't be resurrecting it.

IMO only `LOG_LEVEL_WRN` really makes sense for this macro anyway. `DBG` should always print, `ERR` is fatal, so likely won't happen again anyway, and `INF` is normal operation so shouldn't be spamming the logs in any case. `WRN` is uniquely "user should be notified, the event is unusual, but because its only a warning, may occur many times in quick succession".

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**